Google cast with current library implementation

albaintor opened the pull request:
Same PR as https://github.com/unfoldedcircle/integration-androidtv/pull/56 but with the unmodified library
Thanks to this home assistant library https://github.com/home-assistant/core/tree/dev/homeassistant/components/cast this is now possible to retrieve additional media info and features :

- Media title
- Artist
- Album
- Position
- Duration
- Artwork
- Volume level
- Mute toggle
- Media seeking

Available info :

- Disney+ : media position, duration and title
- AppleTV+ : media position & duration
- Netflix : nothing
- Prime Video : nothing

A setup flow parameter disabled by default lets enable the Chromecast features.
Features are handled dynamically according to this parameter value.

zehnm commented:
I noticed that the friendly app name mapping is no longer working, or gets overwritten with the app id. Probably easy to fix, but I haven't checked yet where it happens. Example where `media_title` is correct, then replaced with the app id:
```
DEBUG:tv:[ShieldTV Wohnzimmer] Update remote with Chromecast info : {<Attributes.MEDIA_TITLE: 'media_title'>: 'Apple TV'}
DEBUG:driver:[ShieldTV Wohnzimmer] device update: {<Attributes.MEDIA_TITLE: 'media_title'>: 'Apple TV'}
DEBUG:ucapi.entities:[configured]: entity 'yyyy' attributes updated
DEBUG:ucapi.api:[('xxxx', 50282)] ->: {"kind": "event", "msg": "entity_change", "msg_data": {"entity_id": "yyyy", "entity_type": "media_player", "attributes": {"media_title": "Apple TV"}}, "cat": "ENTITY"}
DEBUG:tv:[ShieldTV Wohnzimmer] Received Chromecast cast status : CastStatus(is_active_input=False, is_stand_by=True, volume_level=0.46666666865348816, volume_muted=False, app_id='AndroidNativeApp', display_name='Apple TV', namespaces=['urn:x-cast:com.google.cast.media'], session_id='795d0cab-52d1-48f7-a4b1-e700e74f2621', transport_id='795d0cab-52d1-48f7-a4b1-e700e74f2621', status_text='Apple TV', icon_url='', volume_control_type='attenuation')
DEBUG:androidtvremote2:Received: remote_ime_key_inject { app_info { app_package: "com.apple.atve.androidtv.appletv" } }
DEBUG:tv:[ShieldTV Wohnzimmer] current_app: com.apple.atve.androidtv.appletv
DEBUG:driver:[ShieldTV Wohnzimmer] device update: {<Attributes.SOURCE: 'source'>: 'Apple TV', <Attributes.STATE: 'state'>: 'PLAYING', <Attributes.MEDIA_TITLE: 'media_title'>: 'com.apple.atve.androidtv.appletv'}
```
The app id `com.apple.atve.androidtv.appletv` is in the `IdMappings` and works with setting `use_chromecast = false`.

albaintor commented:
> I noticed that the friendly app name mapping is no longer working, or gets overwritten with the app id. Probably easy to fix, but I haven't checked yet where it happens. Example where `media_title` is correct, then replaced with the app id:
> 
> ```
> DEBUG:tv:[ShieldTV Wohnzimmer] Update remote with Chromecast info : {<Attributes.MEDIA_TITLE: 'media_title'>: 'Apple TV'}
> DEBUG:driver:[ShieldTV Wohnzimmer] device update: {<Attributes.MEDIA_TITLE: 'media_title'>: 'Apple TV'}
> DEBUG:ucapi.entities:[configured]: entity 'yyyy' attributes updated
> DEBUG:ucapi.api:[('xxxx', 50282)] ->: {"kind": "event", "msg": "entity_change", "msg_data": {"entity_id": "yyyy", "entity_type": "media_player", "attributes": {"media_title": "Apple TV"}}, "cat": "ENTITY"}
> DEBUG:tv:[ShieldTV Wohnzimmer] Received Chromecast cast status : CastStatus(is_active_input=False, is_stand_by=True, volume_level=0.46666666865348816, volume_muted=False, app_id='AndroidNativeApp', display_name='Apple TV', namespaces=['urn:x-cast:com.google.cast.media'], session_id='795d0cab-52d1-48f7-a4b1-e700e74f2621', transport_id='795d0cab-52d1-48f7-a4b1-e700e74f2621', status_text='Apple TV', icon_url='', volume_control_type='attenuation')
> DEBUG:androidtvremote2:Received: remote_ime_key_inject { app_info { app_package: "com.apple.atve.androidtv.appletv" } }
> DEBUG:tv:[ShieldTV Wohnzimmer] current_app: com.apple.atve.androidtv.appletv
> DEBUG:driver:[ShieldTV Wohnzimmer] device update: {<Attributes.SOURCE: 'source'>: 'Apple TV', <Attributes.STATE: 'state'>: 'PLAYING', <Attributes.MEDIA_TITLE: 'media_title'>: 'com.apple.atve.androidtv.appletv'}
> ```
> 
> The app id `com.apple.atve.androidtv.appletv` is in the `IdMappings` and works with setting `use_chromecast = false`.

Ok I think I know how to fix it

albaintor commented:
This is fixed (app name instead of app id)

albaintor commented:
For now I display in logs this message but it should be commented after testing because too verbose
```python _LOG.debug("[%s] Update from Chromecast info : %s", self.log_id, status)```

zehnm commented on intg-androidtv/tv.py:
```diff
@@ -432,6 +488,28 @@ def _handle_invalid_auth() -> None:
+            if self._chromecast is None:
+                self._chromecast = pychromecast.get_chromecast_from_host(
+                    host=(self._atv.host, None, None, None, None),
+                    tries=10,
```
Why limit retries to 10 only?
According to https://github.com/home-assistant-libs/pychromecast/blob/14.0.7/pychromecast/__init__.py#L296
> Number of retries to perform if the connection fails. None for infinite retries.

Ideally, the library would keep the connection and automatically reconnect forever if set to None. That's my expectation :-)
There doesn't seem much documentation available... So we have to dig through the code.

albaintor replied:
I think that infinite tries is not a good idea : you will have permanent network requests to the device. 
I thought about that : when the device recovers connectivity, there is a callback that I already changed to trigger a reconnection like this :
```python
    def _is_on_updated(self, is_on: bool) -> None:
        """Notify that the Android TV power state is updated."""
        _LOG.info("[%s] is on: %s", self.log_id, is_on)
        update = {}
        if is_on:
            update[MediaAttr.STATE] = media_player.States.ON.value
            # Chromecast service is not accessible when the device is in standby
            try:
                if self._chromecast and not self._chromecast.socket_client.is_alive():
                    self._chromecast.wait(timeout=CONNECTION_TIMEOUT)
            except (RequestTimeout, RuntimeError) as ex:
                _LOG.info("[%s] Chromecast connection error %s", self.log_id, ex)
        else:
            update[MediaAttr.STATE] = media_player.States.OFF.value
        self.events.emit(Events.UPDATE, self._identifier, update)
```
What do you think ?

zehnm commented on intg-androidtv/tv.py:
```diff
@@ -473,6 +551,11 @@ def disconnect(self) -> None:
         self._atv.disconnect()
+        if self._chromecast and self._chromecast.socket_client.is_alive():
+            try:
+                self._chromecast.disconnect(timeout=CONNECTION_TIMEOUT)
```
Is there a reason to wait for disconnection?
Docs:
> Set to 0 to not block.

This can cause issues if we wait here up to 10 seconds.

albaintor replied:
Yes this is better

zehnm reviewed (changes requested):
Some more code-review feedback before I start testing on the Remote

albaintor commented on intg-androidtv/profiles.py:
```diff
@@ -155,7 +179,13 @@ def __init__(self):
                 media_player.Features.SELECT_SOURCE,
+                media_player.Features.SELECT_SOUND_MODE,
                 media_player.Features.MEDIA_TITLE,
+                media_player.Features.MEDIA_ALBUM,
+                media_player.Features.MEDIA_ARTIST,
+                media_player.Features.MEDIA_POSITION,
+                media_player.Features.MEDIA_DURATION,
+                media_player.Features.MEDIA_IMAGE_URL,
```
Yes right, removed

albaintor commented:
Resolved except one thing above around number of retries

albaintor commented:
Hi Markus,

I think I nailed it. I have tested it OK :

1. Start driver => chromecast connection established and mediadata retrieved
2. Disconnect (electrically) the Shield
3. The driver tries to reconnect (infinitely) to the androidtremote service, and tries 10x to reconnect to chromecast
4. Wait until the 10 retries
5. Reconnect the shield
6. The `_is_on_updated` is called when the connection is recovered and the Chromecast instance is reinitialized
7. Media info is updated correctly for chromecast part


The tricky thing was : the chromecast instance cannot be reused after disconnection. I had to flush it and recreate it. Now it works

Also I added a call to connection task in the command wrapper because in some case the state of connection is not correctly detected. In mycase each time I pressed ON or HOME nothing happened.
I added a ` self._loop.create_task(self.connect())` call in the `CannotConnect, ConnectionClosed` exception block
```python
@wraps(func)
    async def wrapper(self: _AndroidTvT, *args: _P.args, **kwargs: _P.kwargs) -> ucapi.StatusCodes:
...
            if (
                not (self._atv and self._atv._remote_message_protocol and self._atv._remote_message_protocol.transport)
                or self._atv._remote_message_protocol.transport.is_closing()
            ):
...
                self._loop.create_task(self.connect())
                return ucapi.StatusCodes.SERVICE_UNAVAILABLE
            return await func(self, *args, **kwargs)
        except (CannotConnect, ConnectionClosed) as ex:
            _LOG.error("[%s] Cannot send command: %s", self.log_id, ex)
            # pylint: disable=W0212
            self._loop.create_task(self.connect())
            return ucapi.StatusCodes.SERVICE_UNAVAILABLE
```

zehnm commented:
I'll update and continue testing. The previous version didn't keep the Cast connection overnight.

albaintor commented:
Yes same here, this should be resolved with the last update

albaintor commented:
Hi Markus,

I have just added a new reconfigure option in the setup flow (only chromecast checkbox at this state)

zehnm commented:
Great, thank you! Less work for me :-)
I did some more testing yesterday and so far it's promising 👍 I will continue tomorrow.

zehnm reviewed (changes requested):
Looks good so far after testing it for a while 👍 In case there are some undetected re-connection issues, we'll just fix it later.
Thanks again for using the official pychromecast library. We can always switch to your asyncio version if it gets merged upstream.

One last small change, then it's good to go: the `SEEK` feature may not be in the default profile, since it's a Google Cast feature.
I'll add the volume controls with the androidtvremote2 library myself after the merge.

albaintor commented on intg-androidtv/profiles.py:
```diff
@@ -168,6 +192,10 @@ def __init__(self):
                 media_player.Features.SUBTITLE,
                 media_player.Features.RECORD,
                 media_player.Features.STOP,
+                media_player.Features.SEEK,
```
Done

albaintor commented on intg-androidtv/profiles.py:
```diff
@@ -168,6 +192,10 @@ def __init__(self):
                 media_player.Features.RECORD,
                 media_player.Features.STOP,
+                media_player.Features.SEEK,
+                media_player.Features.VOLUME,
+                media_player.Features.VOLUME_UP_DOWN,
+                media_player.Features.MUTE_TOGGLE,
```
Ok I let it there

albaintor commented:
Done : removed seek feature from `profiles.py`

zehnm reviewed (approved):
Thanks again! I will merge it and add the alternative volume control if Cast is disabled.

albaintor commented:
You realize that you now have the most advanced controlled implementation of Apple & Android devices !
